design/idempotency_ids.md

```diff
@@ -0,0 +1,103 @@
+
+After reporting a successful commit, clients will send RPCs to the cleaner indicating that this idempotency id is no longer necessary. Once all idempotency ids in a value are no longer necessary, the cleaner may clear that value's key.
+
+This may be a bottleneck and is the main reason why I'm doing disabled by default an intermediate goal.
+
+# Commit protocol
+
+The basic change will be that a commit future will not become ready until the client confirms whether or not the commit succeeded. (`transaction_timed_out` is an unfortunate exception here)
+
+The idempotency id will be automatically added to both the read conflict range and the write conflict range, before makeSelfConflicting is called so that we don't duplicate that work. We can reuse the `\xff/SC/` self-conflicting key space here.
+
+## Did I already commit?
+
+Storage servers would have a new endpoint that clients can use to ask if the transaction for an idempotency id already committed. Clients would need to check every possible shard that their idempotency id may have ended up in.
+
+Storage servers would maintain a map from idempotency id to versionstamp in memory, and clients would need to contact all storage servers responsible for the `[\xff\x02/idmp/, \xff\x02/idmp0)` keyspace to be sure of their commit status. Assuming an idempotency id + versionstamp is 16 + 10 bytes, and that the lifetime of most idempotency ids is less than 1 second, that corresponds to at least 260 MB of memory on the storage server at 1,000,000 transactions/s, which seems acceptable. Let's double that to account for things like hash table load factor and allocating extra memory to ensure amortized constant time insertion. Still seems acceptable. We probably want to use a hashtable with open addressing to avoid frequent heap allocations. I _think_ [swisstables](https://abseil.io/about/design/swisstables) would work here.
+
+When a transaction learns that it did in fact commit, the commit future succeeds, and the versionstamp gets filled with the original, successful transaction's versionstamp. After the successful commit is reported, it's no longer necessary to store its idempotency id. The client will send an rpc to the cleaner role indicating that it can remove this idempotency id.
+
+If a transaction learns that it did in fact _not_ commit, the commit future will fail with an error that indicates that the transaction did not commit. Perhaps `transaction_too_old`.
+
+If a transaction learns that it has been in-flight so long that its idempotency id could have been expired, then it will fail with a new, non-retriable error. It is expected that this will be rare enough that crashing the application is acceptable.
+
+# Considerations
+
+- Additional storage space on the cluster. This can be controlled directly via an idempotency id target bytes knob/config.
+- Potential write hot spot.
+- Cleaner is a potential bottleneck, since all committed transactions need to send an rpc to it.
```

Comment:
Maybe let client tell commit/grv proxiex the idempotency ID (load balancing), which then aggregate these IDs and send to the cleaner (batching).

Reply:
The problem is that we can only be sure that a key can be removed if every idempotency id contained in its value can be removed. This means that either commit proxies would need to propagate ids they hear about to each other, or that all ids for a key are routed to the same commit proxy. In the final design I think the reply to the CommitTransactionRequest will tell the client which proxy to send the id to, so that we can route all ids for a key to the same commit proxy.